"possible bug found" from pool when disk went offline {2.10}{pool}

Hi,

Perhaps these errors from dCache is of interest (likely fix is known, replace disk and restart server):

This happened on an HP DL360 G7 with 4 x D2600 disk shelves behind a P411 controller in SL6.

dcache version 2.10.25

java-1.7.0-openjdk-devel-1.7.0.79-2.5.5.1.el6_6.x86_64
java-1.7.0-openjdk-1.7.0.79-2.5.5.1.el6_6.x86_64

###

A disk went into predictive failure and then the disk (logical drive) was offlined:

<pre>
Jun 22 14:24:03 pacific08 kernel: hpsa 0000:06:00.0: Abort request on C1:B0:T0:L0
Jun 22 14:24:03 pacific08 kernel: hpsa 0000:06:00.0: ABORT REQUEST on C1:B0:T0:L0 Tag:0x00000000:00000090 Command:0x88 SN:0x4c317141  REQUEST SUCCEEDED.
[snip]
Jun 22 22:51:03 pacific08 kernel: hpsa 0000:06:00.0: ABORT REQUEST on C1:B0:T0:L0 Tag:0x00000000:000001f0 Command:0x88 SN:0x4cf98c6f  REQUEST SUCCEEDED.
Jun 22 22:51:33 pacific08 kernel: hpsa 0000:06:00.0: ABORT REQUEST on C1:B0:T0:L0 Tag:0x00000000:000001f0 Command:0x88 SN:0x4cf98c6f  FAILED. Aborted command has not completed after 30 seconds.
Jun 22 22:51:33 pacific08 kernel: hpsa 0000:06:00.0: resetting device 1:0:0:0
Jun 22 22:51:34 pacific08 kernel: hpsa 0000:06:00.0: device is ready.
Jun 22 22:51:34 pacific08 kernel: sd 1:0:0:0: [sdb] Medium access timeout failure. Offlining disk!
Jun 22 22:51:34 pacific08 kernel: sd 1:0:0:0: Device offlined - not ready after error recovery
Jun 22 22:51:34 pacific08 kernel: sd 1:0:0:0: [sdb]  Result: hostbyte=DID_TIME_OUT driverbyte=DRIVER_OK
Jun 22 22:51:34 pacific08 kernel: sd 1:0:0:0: [sdb] CDB: Read(16): 88 00 00 00 00 02 93 5d e3 d8 00 00 00 60 00 00
Jun 22 22:51:34 pacific08 kernel: end_request: I/O error, dev sdb, sector 11062338520
Jun 22 22:51:34 pacific08 kernel: sd 1:0:0:0: rejecting I/O to offline device
Jun 22 22:51:34 pacific08 kernel: sd 1:0:0:0: [sdb] killing request
</pre>

###

Then we saw this from dCache logs: 

<pre>
22 Jun 2015 22:51:35 (csc_fi_29) [] xrootd I/O error: java.io.IOException: Input/output error
22 Jun 2015 22:51:35 (csc_fi_29) [] xrootd I/O error: java.io.IOException: Input/output error
22 Jun 2015 22:51:35 (csc_fi_29) [] xrootd I/O error: java.io.IOException: Input/output error
22 Jun 2015 22:51:35 (csc_fi_29) [] xrootd I/O error: java.io.IOException: Input/output error
22 Jun 2015 22:51:35 (csc_fi_29) [] xrootd I/O error: java.io.IOException: Input/output error
22 Jun 2015 22:51:35 (csc_fi_29) [] xrootd I/O error: java.io.IOException: Input/output error
22 Jun 2015 22:51:35 (csc_fi_29) [] xrootd I/O error: java.io.IOException: Input/output error
22 Jun 2015 22:51:35 (csc_fi_29) [] xrootd I/O error: java.io.IOException: Input/output error
22 Jun 2015 22:51:35 (csc_fi_29) [] xrootd I/O error: java.io.IOException: Input/output error
22 Jun 2015 22:51:35 (csc_fi_29) [] Could not create /srv/dcache/pool1/csc_fi_29/meta/.repository_is_ok
22 Jun 2015 22:51:35 (csc_fi_29) [] Fault occured in repository: I/O test failed
22 Jun 2015 22:51:35 (csc_fi_29) [] Pool mode changed to disabled(fetch,store,stage,p2p-client,p2p-server): Pool disabled: I/O test failed
22 Jun 2015 22:51:35 (csc_fi_29) [door:Xrootd-madhatter@xrootd-madhatterDomain:AAABThy2emc Xrootd-madhatter PoolDeliverFile 000045B1C515108D46A489C671F7A403CC42] PoolIoFileMessage request rej
ected due to disabled(fetch,store,stage,p2p-client,p2p-server)
22 Jun 2015 22:51:35 (csc_fi_29) [] FILE OPEN RESPONSE  connId: -120714739    path :pnfs/csc.fi/data/cms/store/mc/Phys14DR/QCD_Pt-170to300_Tune4C_13TeV_pythia8/GEN-SIM-RAW/AVE20BX25_tsg_casto
r_PHYS14_25_V3-v1/30000/54F82F5D-E2A8-E411-83F8-0025907DC9DC.root,  fileCounter: 90869
22 Jun 2015 22:51:35 (csc_fi_29) [] FILE OPEN RESPONSE  connId: 2143435051    path :pnfs/csc.fi/data/cms/store/mc/Phys14DR/QCD_Pt-170to300_Tune4C_13TeV_pythia8/GEN-SIM-RAW/AVE20BX25_tsg_castor_PHYS14_25_V3-v1/30000/D096402A-CCA8-E411-8466-002590AC505C.root,  fileCounter: 90870
22 Jun 2015 22:51:35 (csc_fi_29) [door:GFTP-pacific01-225012@ftp-pacific01Domain:1435000104292 GFTP-pacific01-225012 PoolAcceptFile 00002083AFC458F843A3AE49131BB2D2B53A] Marking pool entry as BROKEN
22 Jun 2015 22:51:35 (csc_fi_29) [door:GFTP-madhatter-302655@ftp-madhatterDomain:1435002663989 GFTP-madhatter-302655 PoolAcceptFile 0000CE4911CA397A4A3680FE70A1DA7E396A] Possible bug found: No transaction is active
java.lang.IllegalStateException: No transaction is active
        at com.sleepycat.collections.CurrentTransaction.abortTransaction(CurrentTransaction.java:316) ~[je-6.1.5.jar:6.1.5]
        at com.sleepycat.collections.StoredContainer.handleException(StoredContainer.java:456) ~[je-6.1.5.jar:6.1.5]
        at com.sleepycat.collections.StoredContainer.putKeyValue(StoredContainer.java:327) ~[je-6.1.5.jar:6.1.5]
        at com.sleepycat.collections.StoredMap.put(StoredMap.java:279) ~[je-6.1.5.jar:6.1.5]
        at org.dcache.pool.repository.meta.db.CacheRepositoryEntryImpl.setFileAttributes(CacheRepositoryEntryImpl.java:168) ~[dcache-core-2.10.25.jar:2.10.25]
        at org.dcache.pool.repository.v5.CacheRepositoryV5.createEntry(CacheRepositoryV5.java:457) ~[dcache-core-2.10.25.jar:2.10.25]
        at org.dcache.pool.classic.PoolV4.createMover(PoolV4.java:745) [dcache-core-2.10.25.jar:2.10.25]
        at org.dcache.pool.classic.PoolV4.ioFile(PoolV4.java:791) [dcache-core-2.10.25.jar:2.10.25]
        at org.dcache.pool.classic.PoolV4.messageArrived(PoolV4.java:1038) [dcache-core-2.10.25.jar:2.10.25]
        at sun.reflect.GeneratedMethodAccessor44.invoke(Unknown Source) ~[na:na]
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[na:1.7.0_79]
        at java.lang.reflect.Method.invoke(Method.java:606) ~[na:1.7.0_79]
        at org.dcache.cells.CellMessageDispatcher$LongReceiver.deliver(CellMessageDispatcher.java:305) [dcache-core-2.10.25.jar:2.10.25]
        at org.dcache.cells.CellMessageDispatcher.call(CellMessageDispatcher.java:202) [dcache-core-2.10.25.jar:2.10.25]
        at org.dcache.cells.AbstractCell.messageArrived(AbstractCell.java:802) [dcache-core-2.10.25.jar:2.10.25]
        at dmg.cells.nucleus.CellAdapter.messageArrived(CellAdapter.java:860) [cells-2.10.25.jar:2.10.25]
        at dmg.cells.nucleus.CellNucleus$DeliverMessageTask.innerRun(CellNucleus.java:1151) [cells-2.10.25.jar:2.10.25]
        at dmg.cells.nucleus.CellNucleus$AbstractNucleusTask.run(CellNucleus.java:1052) [cells-2.10.25.jar:2.10.25]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145) [na:1.7.0_79]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615) [na:1.7.0_79]
        at java.lang.Thread.run(Thread.java:745) [na:1.7.0_79]
22 Jun 2015 22:51:35 (csc_fi_29) [door:GFTP-pacific01-225121@ftp-pacific01Domain:1435001988125 GFTP-pacific01-225121 PoolAcceptFile 0000C4BED27E55174A2A8614B8409F585B9D] Marking pool entry as BROKEN
22 Jun 2015 22:51:35 (csc_fi_29) [door:GFTP-pacific01-225012@ftp-pacific01Domain:1435000104292 GFTP-pacific01-225012 PoolAcceptFile 00002083AFC458F843A3AE49131BB2D2B53A] Transfer failed in post-processing. Please report this bug to support@dcache.org.
com.sleepycat.je.LogWriteException: (JE 6.1.5) Environment must be closed, caused by: com.sleepycat.je.LogWriteException: Environment invalid because of previous exception: (JE 6.1.5) /srv/dcache/pool1/csc_fi_29/meta IOException during fsync LOG_WRITE: IOException on write, log is likely incomplete. Environment is invalid and must be closed.
        at com.sleepycat.je.LogWriteException.wrapSelf(LogWriteException.java:72) ~[je-6.1.5.jar:6.1.5]
        at com.sleepycat.je.dbi.EnvironmentImpl.checkIfInvalid(EnvironmentImpl.java:1610) ~[je-6.1.5.jar:6.1.5]
        at com.sleepycat.je.Environment.checkEnv(Environment.java:2231) ~[je-6.1.5.jar:6.1.5]
        at com.sleepycat.je.Database.checkEnv(Database.java:2220) ~[je-6.1.5.jar:6.1.5]
        at com.sleepycat.je.Database.openCursor(Database.java:823) ~[je-6.1.5.jar:6.1.5]
        at com.sleepycat.collections.CurrentTransaction.openCursor(CurrentTransaction.java:415) ~[je-6.1.5.jar:6.1.5]
        at com.sleepycat.collections.MyRangeCursor.openCursor(MyRangeCursor.java:53) ~[je-6.1.5.jar:6.1.5]
        at com.sleepycat.collections.MyRangeCursor.<init>(MyRangeCursor.java:29) ~[je-6.1.5.jar:6.1.5]
        at com.sleepycat.collections.DataCursor.init(DataCursor.java:170) ~[je-6.1.5.jar:6.1.5]
        at com.sleepycat.collections.DataCursor.<init>(DataCursor.java:58) ~[je-6.1.5.jar:6.1.5]
        at com.sleepycat.collections.StoredContainer.getValue(StoredContainer.java:300) ~[je-6.1.5.jar:6.1.5]
        at com.sleepycat.collections.StoredMap.get(StoredMap.java:240) ~[je-6.1.5.jar:6.1.5]
        at org.dcache.pool.repository.meta.db.CacheRepositoryEntryImpl.getStorageInfo(CacheRepositoryEntryImpl.java:148) ~[dcache-core-2.10.25.jar:2.10.25]
        at org.dcache.pool.repository.meta.db.CacheRepositoryEntryImpl.getFileAttributes(CacheRepositoryEntryImpl.java:156) ~[dcache-core-2.10.25.jar:2.10.25]
        at org.dcache.pool.repository.v5.CacheEntryImpl.<init>(CacheEntryImpl.java:36) ~[dcache-core-2.10.25.jar:2.10.25]
        at org.dcache.pool.repository.v5.CacheRepositoryV5.setState(CacheRepositoryV5.java:837) ~[dcache-core-2.10.25.jar:2.10.25]
        at org.dcache.pool.repository.v5.WriteHandleImpl.fail(WriteHandleImpl.java:405) ~[dcache-core-2.10.25.jar:2.10.25]
        at org.dcache.pool.repository.v5.WriteHandleImpl.close(WriteHandleImpl.java:418) ~[dcache-core-2.10.25.jar:2.10.25]
        at org.dcache.pool.classic.DefaultPostTransferService$1.run(DefaultPostTransferService.java:108) ~[dcache-core-2.10.25.jar:2.10.25]
        at org.dcache.util.FireAndForgetTask.run(FireAndForgetTask.java:31) [dcache-core-2.10.25.jar:2.10.25]
        at org.dcache.util.CDCExecutorServiceDecorator$1.run(CDCExecutorServiceDecorator.java:104) [dcache-core-2.10.25.jar:2.10.25]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145) [na:1.7.0_79]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615) [na:1.7.0_79]
        at java.lang.Thread.run(Thread.java:745) [na:1.7.0_79]
Caused by: com.sleepycat.je.LogWriteException: Environment invalid because of previous exception: (JE 6.1.5) /srv/dcache/pool1/csc_fi_29/meta IOException during fsync LOG_WRITE: IOException on write, log is likely incomplete. Environment is invalid and must be closed.
        at com.sleepycat.je.log.FileManager.syncLogEnd(FileManager.java:2018) ~[je-6.1.5.jar:6.1.5]
        at com.sleepycat.je.log.FSyncManager.executeFSync(FSyncManager.java:393) ~[je-6.1.5.jar:6.1.5]
        at com.sleepycat.je.log.FSyncManager.sync(FSyncManager.java:343) ~[je-6.1.5.jar:6.1.5]
        at com.sleepycat.je.log.LogManager.multiLog(LogManager.java:329) ~[je-6.1.5.jar:6.1.5]
        at com.sleepycat.je.log.LogManager.log(LogManager.java:273) ~[je-6.1.5.jar:6.1.5]
        at com.sleepycat.je.txn.Txn.logCommitEntry(Txn.java:1004) ~[je-6.1.5.jar:6.1.5]
        at com.sleepycat.je.txn.Txn.commit(Txn.java:761) ~[je-6.1.5.jar:6.1.5]
        at com.sleepycat.je.txn.Txn.commit(Txn.java:620) ~[je-6.1.5.jar:6.1.5]
        at com.sleepycat.je.Transaction.commit(Transaction.java:317) ~[je-6.1.5.jar:6.1.5]
        at com.sleepycat.collections.CurrentTransaction.commitTransaction(CurrentTransaction.java:280) ~[je-6.1.5.jar:6.1.5]
        at com.sleepycat.collections.StoredContainer.commitAutoCommit(StoredContainer.java:448) ~[je-6.1.5.jar:6.1.5]
        at com.sleepycat.collections.StoredContainer.putKeyValue(StoredContainer.java:323) ~[je-6.1.5.jar:6.1.5]
        at com.sleepycat.collections.StoredMap.put(StoredMap.java:279) ~[je-6.1.5.jar:6.1.5]
        at org.dcache.pool.repository.meta.db.CacheRepositoryEntryImpl.setFileAttributes(CacheRepositoryEntryImpl.java:168) ~[dcache-core-2.10.25.jar:2.10.25]
        at org.dcache.pool.repository.v5.CacheRepositoryV5.createEntry(CacheRepositoryV5.java:457) ~[dcache-core-2.10.25.jar:2.10.25]
        at org.dcache.pool.classic.PoolV4.createMover(PoolV4.java:745) ~[dcache-core-2.10.25.jar:2.10.25]
        at org.dcache.pool.classic.PoolV4.ioFile(PoolV4.java:791) ~[dcache-core-2.10.25.jar:2.10.25]
        at org.dcache.pool.classic.PoolV4.messageArrived(PoolV4.java:1038) ~[dcache-core-2.10.25.jar:2.10.25]
        at sun.reflect.GeneratedMethodAccessor44.invoke(Unknown Source) ~[na:na]
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[na:1.7.0_79]
        at java.lang.reflect.Method.invoke(Method.java:606) ~[na:1.7.0_79]
        at org.dcache.cells.CellMessageDispatcher$LongReceiver.deliver(CellMessageDispatcher.java:305) ~[dcache-core-2.10.25.jar:2.10.25]
        at org.dcache.cells.CellMessageDispatcher.call(CellMessageDispatcher.java:202) ~[dcache-core-2.10.25.jar:2.10.25]
        at org.dcache.cells.AbstractCell.messageArrived(AbstractCell.java:802) ~[dcache-core-2.10.25.jar:2.10.25]
        at dmg.cells.nucleus.CellAdapter.messageArrived(CellAdapter.java:860) ~[cells-2.10.25.jar:2.10.25]
        at dmg.cells.nucleus.CellNucleus$DeliverMessageTask.innerRun(CellNucleus.java:1151) ~[cells-2.10.25.jar:2.10.25]
        at dmg.cells.nucleus.CellNucleus$AbstractNucleusTask.run(CellNucleus.java:1052) ~[cells-2.10.25.jar:2.10.25]
        ... 3 common frames omitted
Caused by: java.io.IOException: Input/output error
        at sun.nio.ch.FileDispatcherImpl.force0(Native Method) ~[na:1.7.0_79]
        at sun.nio.ch.FileDispatcherImpl.force(FileDispatcherImpl.java:77) ~[na:1.7.0_79]
        at sun.nio.ch.FileChannelImpl.force(FileChannelImpl.java:377) ~[na:1.7.0_79]
        at com.sleepycat.je.log.FileManager$LogEndFileDescriptor.force(FileManager.java:3011) ~[je-6.1.5.jar:6.1.5]
        at com.sleepycat.je.log.FileManager$LogEndFileDescriptor.access$500(FileManager.java:2678) ~[je-6.1.5.jar:6.1.5]
        at com.sleepycat.je.log.FileManager.syncLogEnd(FileManager.java:2016) ~[je-6.1.5.jar:6.1.5]
        ... 29 common frames omitted
</pre>

--- a/modules/dcache/src/main/java/org/dcache/pool/classic/PoolV4.java
+++ b/modules/dcache/src/main/java/org/dcache/pool/classic/PoolV4.java
@@ -528,7 +528,7 @@ public class PoolV4
             break;
         }
 
-        String message = "Fault occured in " + event.getSource() + ": "
+        String message = "Fault occurred in " + event.getSource() + ": "
                         + event.getMessage() +". " + poolState;
 
         if (cause != null) {

--- a/modules/dcache/src/main/java/org/dcache/pool/repository/ConsistentStore.java
+++ b/modules/dcache/src/main/java/org/dcache/pool/repository/ConsistentStore.java
@@ -118,11 +118,11 @@ public class ConsistentStore
      * redundant meta data entries in the process.
      */
     @Override
-    public synchronized Collection<PnfsId> list()
+    public synchronized Collection<PnfsId> list() throws CacheException
     {
         Collection<PnfsId> files = _fileStore.list();
         Collection<PnfsId> records = _metaDataStore.list();
-        records.removeAll(new HashSet(files));
+        records.removeAll(new HashSet<>(files));
         for (PnfsId id: records) {
             _log.warn(String.format(REMOVING_REDUNDANT_META_DATA, id));
             _metaDataStore.remove(id);
@@ -202,7 +202,7 @@ public class ConsistentStore
         return entry;
     }
 
-    private boolean isBroken(MetaDataRecord entry)
+    private boolean isBroken(MetaDataRecord entry) throws CacheException
     {
         boolean isBroken = true;
         if (entry != null) {
@@ -402,7 +402,7 @@ public class ConsistentStore
      * Calls through to the wrapped meta data store.
      */
     @Override
-    public void remove(PnfsId id)
+    public void remove(PnfsId id) throws CacheException
     {
         File f = _fileStore.get(id);
         if (!f.delete() && f.exists()) {
@@ -453,7 +453,7 @@ public class ConsistentStore
         return _metaDataStore.getTotalSpace();
     }
 
-    private void delete(PnfsId id, File file)
+    private void delete(PnfsId id, File file) throws CacheException
     {
         _metaDataStore.remove(id);
         if (!file.delete() && file.exists()) {

--- a/modules/dcache/src/main/java/org/dcache/pool/repository/MetaDataCache.java
+++ b/modules/dcache/src/main/java/org/dcache/pool/repository/MetaDataCache.java
@@ -40,7 +40,7 @@ public class MetaDataCache
      *
      * The operation may be slow as the list method of inner is called.
      */
-    public MetaDataCache(MetaDataStore inner)
+    public MetaDataCache(MetaDataStore inner) throws CacheException
     {
         _inner = inner;
 
@@ -145,7 +145,7 @@ public class MetaDataCache
             return _record;
         }
 
-        private synchronized void remove()
+        private synchronized void remove() throws CacheException
         {
             if (_entries.get(_id) == this) {
                 assert _entries.get(_id) == this;
@@ -181,7 +181,7 @@ public class MetaDataCache
     }
 
     @Override
-    public void remove(PnfsId id)
+    public void remove(PnfsId id) throws CacheException
     {
         Monitor monitor = _entries.get(id);
         if (monitor != null) {

--- a/modules/dcache/src/main/java/org/dcache/pool/repository/MetaDataRecord.java
+++ b/modules/dcache/src/main/java/org/dcache/pool/repository/MetaDataRecord.java
@@ -4,6 +4,7 @@ import java.io.File;
 import java.util.Collection;
 
 import diskCacheV111.util.CacheException;
+import diskCacheV111.util.DiskErrorCacheException;
 import diskCacheV111.util.PnfsId;
 
 import org.dcache.vehicles.FileAttributes;
@@ -44,7 +45,7 @@ public interface MetaDataRecord
 
     public void setFileAttributes(FileAttributes attributes) throws CacheException;
 
-    public FileAttributes getFileAttributes();
+    public FileAttributes getFileAttributes() throws CacheException;
 
     public void setState(EntryState state)
         throws CacheException;
@@ -83,7 +84,7 @@ public interface MetaDataRecord
      *
      * @return The expired sticky flags removed from the record.
      */
-    public Collection<StickyRecord> removeExpiredStickyFlags();
+    public Collection<StickyRecord> removeExpiredStickyFlags() throws CacheException;
 
     /**
      * Set sticky flag for a given owner and time. There is at most

--- a/modules/dcache/src/main/java/org/dcache/pool/repository/MetaDataStore.java
+++ b/modules/dcache/src/main/java/org/dcache/pool/repository/MetaDataStore.java
@@ -18,7 +18,7 @@ public interface MetaDataStore
     /**
      * Returns a collection of PNFS ids of available entries.
      */
-    Collection<PnfsId> list();
+    Collection<PnfsId> list() throws CacheException;
 
     /**
      * Retrieves an existing entry previously created with
@@ -79,7 +79,8 @@ public interface MetaDataStore
      *
      * @param id PNFS id of the entry to return.
      */
-    void remove(PnfsId id);
+    void remove(PnfsId id)
+            throws CacheException;
 
     /**
      * Returns whether the store appears healthy. How this is

--- a/modules/dcache/src/main/java/org/dcache/pool/repository/Repository.java
+++ b/modules/dcache/src/main/java/org/dcache/pool/repository/Repository.java
@@ -37,7 +37,7 @@ public interface Repository
      * @throws IllegalStateException if called multiple times
      */
     void init()
-        throws IllegalStateException;
+            throws IllegalStateException, CacheException;
 
     /**
      * Loads the repository from the on disk state. Must be done

--- a/modules/dcache/src/main/java/org/dcache/pool/repository/meta/db/BerkeleyDBMetaDataRepository.java
+++ b/modules/dcache/src/main/java/org/dcache/pool/repository/meta/db/BerkeleyDBMetaDataRepository.java
@@ -10,11 +10,13 @@ import org.slf4j.LoggerFactory;
 import java.io.File;
 import java.io.FileNotFoundException;
 import java.io.IOException;
+import java.nio.file.Files;
 import java.util.Collection;
 import java.util.HashSet;
 import java.util.Set;
 
 import diskCacheV111.util.CacheException;
+import diskCacheV111.util.DiskErrorCacheException;
 import diskCacheV111.util.PnfsId;
 import diskCacheV111.vehicles.StorageInfo;
 
@@ -101,22 +103,30 @@ public class BerkeleyDBMetaDataRepository
     }
 
     @Override
-    public Collection<PnfsId> list()
+    public Collection<PnfsId> list() throws DiskErrorCacheException
     {
-        Set<PnfsId> ids = new HashSet<>();
-        for (Object id: _views.getStorageInfoMap().keySet()) {
-            ids.add(new PnfsId((String) id));
+        try {
+            Set<PnfsId> ids = new HashSet<>();
+            for (Object id: _views.getStorageInfoMap().keySet()) {
+                ids.add(new PnfsId((String) id));
+            }
+            for (Object id: _views.getStateMap().keySet()) {
+                ids.add(new PnfsId((String) id));
+            }
+            return ids;
+        } catch (DatabaseException e) {
+            throw new DiskErrorCacheException("Meta data lookup failed: " + e.getMessage(), e);
         }
-        for (Object id: _views.getStateMap().keySet()) {
-            ids.add(new PnfsId((String) id));
-        }
-        return ids;
     }
 
     @Override
-    public MetaDataRecord get(PnfsId id)
+    public MetaDataRecord get(PnfsId id) throws DiskErrorCacheException
     {
-        return CacheRepositoryEntryImpl.load(this, id);
+        try {
+            return CacheRepositoryEntryImpl.load(this, id);
+        } catch (DatabaseException e) {
+            throw new DiskErrorCacheException("Meta data updated failed: " + e.getMessage(), e);
+        }
     }
 
     /**
@@ -124,7 +134,7 @@ public class BerkeleyDBMetaDataRepository
      */
     @Override
     public MetaDataRecord create(PnfsId id)
-        throws DuplicateEntryException
+            throws DuplicateEntryException, DiskErrorCacheException
     {
         /* CacheRepositoryEntryImpl.load silently drops incomplete
          * entries. To conform to the contract of the MetaDataStore
@@ -154,10 +164,14 @@ public class BerkeleyDBMetaDataRepository
     }
 
     @Override
-    public void remove(PnfsId id)
+    public void remove(PnfsId id) throws DiskErrorCacheException
     {
-        _views.getStorageInfoMap().remove(id.toString());
-        _views.getStateMap().remove(id.toString());
+        try {
+            _views.getStorageInfoMap().remove(id.toString());
+            _views.getStateMap().remove(id.toString());
+        } catch (DatabaseException e) {
+            throw new DiskErrorCacheException("Meta data update failed: " + e.getMessage(), e);
+        }
     }
 
     @Override
@@ -165,7 +179,7 @@ public class BerkeleyDBMetaDataRepository
     {
         File tmp = new File(_dir, ".repository_is_ok");
         try {
-            tmp.delete();
+            Files.deleteIfExists(tmp.toPath());
             tmp.deleteOnExit();
 
             if (!tmp.createNewFile() || !tmp.exists()) {
@@ -178,10 +192,10 @@ public class BerkeleyDBMetaDataRepository
             }
 
             return true;
-	} catch (IOException e) {
+        } catch (IOException e) {
             _log.error("Failed to touch " + tmp + ": " + e.getMessage());
             return false;
-	}
+        }
     }
 
     /**

--- a/modules/dcache/src/main/java/org/dcache/pool/repository/meta/db/CacheRepositoryEntryImpl.java
+++ b/modules/dcache/src/main/java/org/dcache/pool/repository/meta/db/CacheRepositoryEntryImpl.java
@@ -2,6 +2,8 @@ package org.dcache.pool.repository.meta.db;
 
 import com.google.common.collect.ImmutableList;
 import com.google.common.collect.Lists;
+import com.google.common.io.Files;
+import com.sleepycat.je.DatabaseException;
 import com.sleepycat.util.RuntimeExceptionWrapper;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
@@ -71,7 +73,7 @@ public class CacheRepositoryEntryImpl implements MetaDataRecord
     }
 
     public CacheRepositoryEntryImpl(BerkeleyDBMetaDataRepository repository,
-                                    MetaDataRecord entry)
+                                    MetaDataRecord entry) throws CacheException
     {
         _repository   = repository;
         _pnfsId       = entry.getPnfsId();
@@ -183,25 +185,33 @@ public class CacheRepositoryEntryImpl implements MetaDataRecord
     }
 
     @Override
-    public synchronized FileAttributes getFileAttributes()
+    public synchronized FileAttributes getFileAttributes() throws DiskErrorCacheException
     {
-        FileAttributes attributes = new FileAttributes();
-        attributes.setPnfsId(_pnfsId);
-        StorageInfo storageInfo = getStorageInfo();
-        if (storageInfo != null) {
-            StorageInfos.injectInto(storageInfo, attributes);
+        try {
+            FileAttributes attributes = new FileAttributes();
+            attributes.setPnfsId(_pnfsId);
+            StorageInfo storageInfo = getStorageInfo();
+            if (storageInfo != null) {
+                StorageInfos.injectInto(storageInfo, attributes);
+            }
+            return attributes;
+        } catch (DatabaseException e) {
+            throw new DiskErrorCacheException("Meta data lookup failed: " + e.getMessage(), e);
         }
-        return attributes;
     }
 
     @Override
-    public void setFileAttributes(FileAttributes attributes)
+    public void setFileAttributes(FileAttributes attributes) throws DiskErrorCacheException
     {
-        String id = _pnfsId.toString();
-        if (attributes.isDefined(FileAttribute.STORAGEINFO)) {
-            _repository.getStorageInfoMap().put(id, StorageInfos.extractFrom(attributes));
-        } else {
-            _repository.getStorageInfoMap().remove(id);
+        try {
+            String id = _pnfsId.toString();
+            if (attributes.isDefined(FileAttribute.STORAGEINFO)) {
+                _repository.getStorageInfoMap().put(id, StorageInfos.extractFrom(attributes));
+            } else {
+                _repository.getStorageInfoMap().remove(id);
+            }
+        } catch (DatabaseException e) {
+            throw new DiskErrorCacheException("Meta data update failed: " + e.getMessage(), e);
         }
     }
 
@@ -218,7 +228,7 @@ public class CacheRepositoryEntryImpl implements MetaDataRecord
     }
 
     @Override
-    public synchronized void setState(EntryState state)
+    public synchronized void setState(EntryState state) throws DiskErrorCacheException
     {
         if (_state != state) {
             _state = state;
@@ -239,7 +249,7 @@ public class CacheRepositoryEntryImpl implements MetaDataRecord
     }
 
     @Override
-    public synchronized Collection<StickyRecord> removeExpiredStickyFlags()
+    public synchronized Collection<StickyRecord> removeExpiredStickyFlags() throws DiskErrorCacheException
     {
         long now = System.currentTimeMillis();
         List<StickyRecord> removed = Lists.newArrayList(filter(_sticky, r -> !r.isValidAt(now)));
@@ -287,9 +297,13 @@ public class CacheRepositoryEntryImpl implements MetaDataRecord
         return _sticky;
     }
 
-    private synchronized void storeState()
+    private synchronized void storeState() throws DiskErrorCacheException
     {
-        _repository.getStateMap().put(_pnfsId.toString(), new CacheRepositoryEntryState(_state, _sticky));
+        try {
+            _repository.getStateMap().put(_pnfsId.toString(), new CacheRepositoryEntryState(_state, _sticky));
+        } catch (DatabaseException e) {
+            throw new DiskErrorCacheException("Meta data updated failed: " + e.getMessage(), e);
+        }
     }
 
     static CacheRepositoryEntryImpl load(BerkeleyDBMetaDataRepository repository,

--- a/modules/dcache/src/main/java/org/dcache/pool/repository/meta/file/CacheRepositoryEntryImpl.java
+++ b/modules/dcache/src/main/java/org/dcache/pool/repository/meta/file/CacheRepositoryEntryImpl.java
@@ -11,6 +11,10 @@ import java.io.FileOutputStream;
 import java.io.IOException;
 import java.io.ObjectInputStream;
 import java.io.ObjectOutputStream;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.nio.file.StandardCopyOption;
 import java.util.Collection;
 
 import diskCacheV111.util.CacheException;
@@ -185,14 +189,18 @@ public class CacheRepositoryEntryImpl implements MetaDataRecord
         try {
             _state.setState(state);
         } catch (IOException e) {
-            throw new CacheException(e.getMessage());
+            throw new DiskErrorCacheException(e.getMessage(), e);
         }
     }
 
     @Override
-    public synchronized Collection<StickyRecord> removeExpiredStickyFlags()
+    public synchronized Collection<StickyRecord> removeExpiredStickyFlags() throws CacheException
     {
-        return _state.removeExpiredStickyFlags();
+        try {
+            return _state.removeExpiredStickyFlags();
+        } catch (IOException e) {
+            throw new DiskErrorCacheException(e.getMessage(), e);
+        }
     }
 
     @Override
@@ -204,7 +212,7 @@ public class CacheRepositoryEntryImpl implements MetaDataRecord
             return false;
 
         } catch (IllegalStateException | IOException e) {
-            throw new CacheException(e.getMessage());
+            throw new DiskErrorCacheException(e.getMessage(), e);
         }
     }
 
@@ -221,10 +229,14 @@ public class CacheRepositoryEntryImpl implements MetaDataRecord
 
     @Override
     public void setFileAttributes(FileAttributes attributes) throws CacheException {
-        if (attributes.isDefined(FileAttribute.STORAGEINFO)) {
-            setStorageInfo(StorageInfos.extractFrom(attributes));
-        } else {
-            setStorageInfo(null);
+        try {
+            if (attributes.isDefined(FileAttribute.STORAGEINFO)) {
+                setStorageInfo(StorageInfos.extractFrom(attributes));
+            } else {
+                setStorageInfo(null);
+            }
+        } catch (IOException e) {
+            throw new DiskErrorCacheException(_pnfsId + " " + e.getMessage(), e);
         }
     }
 
@@ -236,8 +248,7 @@ public class CacheRepositoryEntryImpl implements MetaDataRecord
                 _dataFile.createNewFile();
             }
         }catch(IOException ee){
-            throw new
-                CacheException("Io Error creating : "+_dataFile ) ;
+            throw new DiskErrorCacheException("Io Error creating : "+_dataFile ,ee);
         }
         setLastAccessTime(System.currentTimeMillis());
     }
@@ -247,25 +258,16 @@ public class CacheRepositoryEntryImpl implements MetaDataRecord
         return _storageInfo;
     }
 
-    private synchronized void setStorageInfo(StorageInfo storageInfo) throws CacheException {
+    private synchronized void setStorageInfo(StorageInfo storageInfo) throws IOException
+    {
+        File siFileTemp = File.createTempFile(_siFile.getName(), null, _siFile.getParentFile());
         try {
-            File siFileTemp = File.createTempFile(_siFile.getName(), null, _siFile.getParentFile());
-            try {
-                try (ObjectOutputStream objectOut = new ObjectOutputStream(new FileOutputStream(siFileTemp))) {
-                    objectOut.writeObject(storageInfo);
-                }
-                if (!siFileTemp.renameTo(_siFile)) {
-                    // TODO: disk io error code here
-                    throw new CacheException(10, _pnfsId + " rename failed");
-                }
-            } finally {
-                if (siFileTemp.exists()) {
-                    siFileTemp.delete();
-                }
+            try (ObjectOutputStream objectOut = new ObjectOutputStream(new FileOutputStream(siFileTemp))) {
+                objectOut.writeObject(storageInfo);
             }
-        } catch (IOException e) {
-            // TODO: disk io error code here
-            throw new CacheException(10, _pnfsId + " " + e.getMessage(), e);
+            Files.move(siFileTemp.toPath(), _siFile.toPath(), StandardCopyOption.ATOMIC_MOVE, StandardCopyOption.REPLACE_EXISTING);
+        } finally {
+            Files.deleteIfExists(siFileTemp.toPath());
         }
         _storageInfo = storageInfo;
     }

--- a/modules/dcache/src/main/java/org/dcache/pool/repository/meta/file/FileMetaDataRepository.java
+++ b/modules/dcache/src/main/java/org/dcache/pool/repository/meta/file/FileMetaDataRepository.java
@@ -6,6 +6,7 @@ import org.slf4j.LoggerFactory;
 import java.io.File;
 import java.io.FileNotFoundException;
 import java.io.IOException;
+import java.nio.file.Files;
 import java.util.Collection;
 import java.util.HashSet;
 import java.util.Set;
@@ -157,19 +158,24 @@ public class FileMetaDataRepository
         } catch (IOException e) {
             throw new DiskErrorCacheException(
                     "Failed to read meta data for " + id + ": " + e.getMessage(), e);
-        } catch (RuntimeException e) {
-            throw new RuntimeException("Failed to read meta data for " + id, e);
         }
         return null;
     }
 
     @Override
-    public void remove(PnfsId id) {
-        File controlFile = new File(_metadir, id.toString());
-        File siFile = new File(_metadir, "SI-"+id.toString());
+    public void remove(PnfsId id)
+            throws CacheException
+    {
+        try {
+            File controlFile = new File(_metadir, id.toString());
+            File siFile = new File(_metadir, "SI-"+id.toString());
 
-        controlFile.delete();
-        siFile.delete();
+            Files.deleteIfExists(controlFile.toPath());
+            Files.deleteIfExists(siFile.toPath());
+        } catch (IOException e) {
+            throw new DiskErrorCacheException(
+                    "Failed to remove meta data for " + id + ": " + e.getMessage(), e);
+        }
     }
 
     @Override

--- a/modules/dcache/src/main/java/org/dcache/pool/repository/v3/entry/CacheRepositoryEntryState.java
+++ b/modules/dcache/src/main/java/org/dcache/pool/repository/v3/entry/CacheRepositoryEntryState.java
@@ -69,15 +69,11 @@ public class CacheRepositoryEntryState
         makeStatePersistent();
     }
 
-    public List<StickyRecord> removeExpiredStickyFlags()
+    public List<StickyRecord> removeExpiredStickyFlags() throws IOException
     {
         List<StickyRecord> removed = _sticky.removeExpired();
-        try {
-            if (!removed.isEmpty()) {
-                makeStatePersistent();
-            }
-        } catch (IOException e) {
-            _log.error("Failed to store repository state: {}", e.getMessage());
+        if (!removed.isEmpty()) {
+            makeStatePersistent();
         }
         return removed;
     }

--- a/modules/dcache/src/main/java/org/dcache/pool/repository/v5/CacheEntryImpl.java
+++ b/modules/dcache/src/main/java/org/dcache/pool/repository/v5/CacheEntryImpl.java
@@ -2,6 +2,8 @@ package org.dcache.pool.repository.v5;
 
 import java.util.Collection;
 
+import diskCacheV111.util.CacheException;
+import diskCacheV111.util.DiskErrorCacheException;
 import diskCacheV111.util.PnfsId;
 import diskCacheV111.vehicles.StorageInfo;
 
@@ -23,7 +25,7 @@ public class CacheEntryImpl implements CacheEntry
     private final Collection<StickyRecord> _sticky;
     private final FileAttributes _fileAttributes;
 
-    public CacheEntryImpl(MetaDataRecord entry)
+    public CacheEntryImpl(MetaDataRecord entry) throws CacheException
     {
         synchronized (entry) {
             _size = entry.getSize();

--- a/modules/dcache/src/main/java/org/dcache/pool/repository/v5/CacheRepositoryV5.java
+++ b/modules/dcache/src/main/java/org/dcache/pool/repository/v5/CacheRepositoryV5.java
@@ -1,5 +1,6 @@
 package org.dcache.pool.repository.v5;
 
+import com.google.common.base.Throwables;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
@@ -301,7 +302,7 @@ public class CacheRepositoryV5
 
     @Override
     public void init()
-        throws IllegalStateException
+            throws IllegalStateException, CacheException
     {
         synchronized (this) {
             assert _pnfs != null : "Pnfs handler must be set";
@@ -423,7 +424,15 @@ public class CacheRepositoryV5
     public Iterator<PnfsId> iterator()
     {
         assertOpen();
-        return Collections.unmodifiableCollection(_store.list()).iterator();
+        try {
+            return Collections.unmodifiableCollection(_store.list()).iterator();
+        } catch (DiskErrorCacheException | RuntimeException e) {
+            fail(FaultAction.DEAD, "Internal repository error", e);
+            throw Throwables.propagate(e);
+        } catch (CacheException e) {
+            fail(FaultAction.READONLY, "Internal repository error", e);
+            throw Throwables.propagate(e);
+        }
     }
 
     @Override
@@ -477,6 +486,12 @@ public class CacheRepositoryV5
              */
             _pnfs.notify(new PnfsAddCacheLocationMessage(id, getPoolName()));
             throw new FileInCacheException("Entry already exists: " + id);
+        } catch (RuntimeException e) {
+            fail(FaultAction.DEAD, "Internal repository error", e);
+            throw e;
+        } catch (DiskErrorCacheException e) {
+            fail(FaultAction.DEAD, "Internal repository error", e);
+            throw new RuntimeException("Internal repository error", e);
         } catch (CacheException e) {
             fail(FaultAction.READONLY, "Internal repository error", e);
             throw new RuntimeException("Internal repository error", e);
@@ -534,16 +549,18 @@ public class CacheRepositoryV5
             /* Somebody got the idea that we have the file, so we make
              * sure to remove any stray pointers.
              */
-            MetaDataRecord entry;
             try {
-                entry = _store.create(id);
+                MetaDataRecord entry = _store.create(id);
+                setState(entry, REMOVED);
             } catch (DuplicateEntryException concurrentCreation) {
                 return openEntry(id, flags);
+            } catch (DiskErrorCacheException | RuntimeException f) {
+                fail(FaultAction.DEAD, "Internal repository error", f);
+                e.addSuppressed(f);
             }
-            setState(entry, REMOVED);
             throw e;
-        } catch (DiskErrorCacheException e) {
-            fail(FaultAction.READONLY, "Internal repository error", e);
+        } catch (DiskErrorCacheException | RuntimeException e) {
+            fail(FaultAction.DEAD, "Internal repository error", e);
             throw e;
         }
     }
@@ -562,8 +579,8 @@ public class CacheRepositoryV5
                 }
                 return new CacheEntryImpl(entry);
             }
-        } catch (DiskErrorCacheException e) {
-            fail(FaultAction.READONLY, "Internal repository error", e);
+        } catch (RuntimeException | DiskErrorCacheException e) {
+            fail(FaultAction.DEAD, "Internal repository error", e);
             throw e;
         }
     }
@@ -590,11 +607,17 @@ public class CacheRepositoryV5
              */
             try {
                 entry = _store.create(id);
+                setState(entry, REMOVED);
             } catch (DuplicateEntryException concurrentCreation) {
                 setSticky(id, owner, expire, overwrite);
                 return;
+            } catch (DiskErrorCacheException | RuntimeException f) {
+                fail(FaultAction.DEAD, "Internal repository error", f);
+                e.addSuppressed(f);
             }
-            setState(entry, REMOVED);
+            throw e;
+        } catch (DiskErrorCacheException | RuntimeException e) {
+            fail(FaultAction.DEAD, "Internal repository error", e);
             throw e;
         }
 
@@ -689,8 +712,8 @@ public class CacheRepositoryV5
             if (state != REMOVED) {
                 throw new IllegalTransitionException(id, NEW, state);
             }
-        } catch (DiskErrorCacheException e) {
-            fail(FaultAction.READONLY, "Internal repository error", e);
+        } catch (RuntimeException | DiskErrorCacheException e) {
+            fail(FaultAction.DEAD, "Internal repository error", e);
             throw e;
         }
     }
@@ -740,6 +763,9 @@ public class CacheRepositoryV5
             return getMetaDataRecord(id).getState();
         } catch (FileNotInCacheException e) {
             return NEW;
+        } catch (DiskErrorCacheException | RuntimeException e) {
+            fail(FaultAction.DEAD, "Internal repository error", e);
+            throw e;
         }
     }
 
@@ -747,7 +773,11 @@ public class CacheRepositoryV5
     public void getInfo(PrintWriter pw)
     {
         pw.println("State : " + _state);
-        pw.println("Files : " + _store.list().size());
+        try {
+            pw.println("Files : " + _store.list().size());
+        } catch (CacheException e) {
+            pw.println("Files : " + e.getMessage());
+        }
 
         SpaceRecord space = getSpaceRecord();
         long total = space.getTotalSpace();
@@ -888,8 +918,8 @@ public class CacheRepositoryV5
         } catch (CacheException e) {
             fail(FaultAction.READONLY, "Internal repository error", e);
             throw new RuntimeException("Internal repository error", e);
-        } catch (Error | RuntimeException e) {
-            fail(FaultAction.DEAD, "Internal repository error.", e);
+        } catch (RuntimeException e) {
+            fail(FaultAction.DEAD, "Internal repository error", e);
             throw e;
         }
     }
@@ -919,8 +949,8 @@ public class CacheRepositoryV5
         } catch (CacheException e) {
             fail(FaultAction.READONLY, "Internal repository error", e);
             throw new RuntimeException("Internal repository error", e);
-        } catch (Error | RuntimeException e) {
-            fail(FaultAction.DEAD, "Internal repository error.", e);
+        } catch (RuntimeException e) {
+            fail(FaultAction.DEAD, "Internal repository error", e);
             throw e;
         }
     }
@@ -974,29 +1004,35 @@ public class CacheRepositoryV5
      */
     void destroyWhenRemovedAndUnused(MetaDataRecord entry)
     {
-        synchronized (entry) {
-            EntryState state = entry.getState();
-            PnfsId id = entry.getPnfsId();
-            if (entry.getLinkCount() == 0 && state == EntryState.REMOVED) {
-                /* Setting the entry to DESTROYED ensures that we only deallocate it once.
-                 */
-                setState(entry, DESTROYED);
-                _store.remove(id);
+        try {
+            synchronized (entry) {
+                EntryState state = entry.getState();
+                PnfsId id = entry.getPnfsId();
+                if (entry.getLinkCount() == 0 && state == EntryState.REMOVED) {
+                    /* Setting the entry to DESTROYED ensures that we only deallocate it once.
+                     */
+                    setState(entry, DESTROYED);
+                    _store.remove(id);
 
-                /* It is essential to free after we removed the file: This is the opposite
-                 * of what happens during allocation, in which we allocate before writing
-                 * to disk. We rely on never having anything on disk that we haven't accounted
-                 * for in the Account object.
-                 */
-                _account.free(entry.getSize());
+                    /* It is essential to free after we removed the file: This is the opposite
+                     * of what happens during allocation, in which we allocate before writing
+                     * to disk. We rely on never having anything on disk that we haven't accounted
+                     * for in the Account object.
+                     */
+                    _account.free(entry.getSize());
+                }
             }
+        } catch (DiskErrorCacheException | RuntimeException e) {
+            fail(FaultAction.DEAD, "Internal repository error", e);
+        } catch (CacheException e) {
+            fail(FaultAction.READONLY, "Internal repository error", e);
         }
     }
 
     /**
      * Removes all expired sticky flags of entry.
      */
-    private void removeExpiredStickyFlags(MetaDataRecord entry)
+    private void removeExpiredStickyFlags(MetaDataRecord entry) throws CacheException
     {
         synchronized (entry) {
             CacheEntry oldEntry = new CacheEntryImpl(entry);
@@ -1082,8 +1118,12 @@ public class CacheRepositoryV5
         @Override
         public void run()
         {
-            _tasks.remove(_entry.getPnfsId());
-            removeExpiredStickyFlags(_entry);
+            try {
+                _tasks.remove(_entry.getPnfsId());
+                removeExpiredStickyFlags(_entry);
+            } catch (CacheException | RuntimeException e) {
+                fail(FaultAction.DEAD, "Internal repository error", e);
+            }
         }
     }
 

--- a/modules/dcache/src/main/java/org/dcache/pool/repository/v5/ReadHandleImpl.java
+++ b/modules/dcache/src/main/java/org/dcache/pool/repository/v5/ReadHandleImpl.java
@@ -5,6 +5,7 @@ import java.util.EnumSet;
 import java.util.Set;
 
 import diskCacheV111.util.CacheException;
+import diskCacheV111.util.DiskErrorCacheException;
 import diskCacheV111.util.PnfsHandler;
 
 import org.dcache.namespace.FileAttribute;
@@ -26,7 +27,7 @@ class ReadHandleImpl implements ReplicaDescriptor
 
     ReadHandleImpl(CacheRepositoryV5 repository,
                    PnfsHandler pnfs,
-                   MetaDataRecord entry)
+                   MetaDataRecord entry) throws CacheException
     {
         _repository = checkNotNull(repository);
         _pnfs = checkNotNull(pnfs);

--- a/modules/dcache/src/test/java/org/dcache/pool/repository/ConsistentStoreTest.java
+++ b/modules/dcache/src/test/java/org/dcache/pool/repository/ConsistentStoreTest.java
@@ -333,7 +333,7 @@ public class ConsistentStoreTest
     }
 
     @Test
-    public void shouldSilentlyIgnoreRemoveOfNonExistingReplicas()
+    public void shouldSilentlyIgnoreRemoveOfNonExistingReplicas() throws CacheException
     {
         _consistentStore.remove(PNFSID);
     }

